### ISSUE TITLE
Add regression tests for Vanilla-vs-GRN divergence/convergence on Charbonnier loss

### ISSUE THREAD

**Issue by sachncs** [sachncs/gradient-regularized-newton-boosting-trees#16]
## Problem

The paper's central empirical claim (Section 6, Figure 1) is that **Vanilla Restricted Newton boosting diverges on the Charbonnier loss** while **GRN converges with the adaptive λ_k**. This is the headline result of the entire codebase.

The repository's tests do not cover this divergence behavior:

- `tests/test_boosting.py:test_grn_adaptive_lambda_increases` only asserts `any(lam > 0 for lam in lam_history)` — it does NOT check that GRN converges.
- `tests/test_boosting.py:test_loss_decreases_for_strongly_convex` asserts MSE loss decreases, which is the trivial strongly-convex case — not the interesting claim.
- `experiments/wine_charbonnier.py` and `experiments/ablations.py` are run in CI as smoke tests, but their numerical outputs are not asserted.

If a refactor accidentally changes the GRN algorithm to diverge, no CI check would catch it.

## Evidence

- `tests/test_boosting.py:42-59` — GRN adaptive lambda test is qualitative only.
- `tests/test_boosting.py:62-79` — loss-decrease test uses MSE (strongly convex; trivial).
- Manual reproduction confirms Vanilla diverges (loss 2.28 → 1.89×10¹⁰ over 50 rounds, η=1, lam_base=0). But this is not under regression test.

## Root Cause

The smoke-test experiments are run, but no baseline thresholds are asserted. The qualitative behavior of the paper's headline claim is therefore not under CI.

## Impact

- **Critical regression risk**: a refactor that swaps the order of operations, changes the sign of an update, or otherwise breaks the algorithm will not be caught by CI.
- The CI experiment smoke tests are essentially no-ops from a regression standpoint.

## Recommendation

Add three regression tests to `tests/test_boosting.py`:

```python
def test_vanilla_diverges_on_charbonnier(synthetic_regression):
    """Vanilla Newton must diverge on Charbonnier with lam_base=0 and eta=1."""
    x, y = synthetic_regression
    model = VanillaNewtonBoosting(
        loss=CharbonnierLoss(),
        n_estimators=50,
        learning_rate=1.0,
        max_depth=4,
        lam_base=0.0,
    )
    model.fit(x, y)
    losses = model.history.get("loss")
    assert losses[-1] > losses[0] * 10, (
        f"Vanilla Newton expected to diverge; got {losses[0]:.2f} -> {losses[-1]:.2f}"
    )


def test_grn_converges_on_charbonnier(synthetic_regression):
    """GRN must converge (loss decreases) on Charbonnier."""
    x, y = synthetic_regression
    model = GradientRegularizedNewtonBoosting(
        loss=CharbonnierLoss(),
        n_estimators=50,
        learning_rate=1.0,
        max_depth=4,
        lam_base=0.0,
    )
    model.fit(x, y)
    losses = model.history.get("loss")
    assert losses[-1] < losses[0], (
        f"GRN expected to decrease loss; got {losses[0]:.4f} -> {losses[-1]:.4f}"
    )


def test_static_high_lambda_biased(synthetic_regression):
    """Static high lam_base plateaus above the GRN limit."""
    x, y = synthetic_regression
    grn = GradientRegularizedNewtonBoosting(
        loss=CharbonnierLoss(),
        n_estimators=100,
        learning_rate=1.0,
        max_depth=4,
        lam_base=0.0,
    )
    grn.fit(x, y)
    static = VanillaNewtonBoosting(
        loss=CharbonnierLoss(),
        n_estimators=100,
        learning_rate=1.0,
        max_depth=4,
        lam_base=10.0,
    )
    static.fit(x, y)
    assert static.history.get("loss")[-1] > grn.history.get("loss")[-1]
```

## Acceptance Criteria

- [ ] Three new tests added to `tests/test_boosting.py`.
- [ ] All three pass on the current implementation.
- [ ] A deliberate fault injection (e.g., flip a sign in `compute_lambda`) causes at least one test to fail.

## Verification

```bash
pytest tests/test_boosting.py -v -k "charbonnier or diverges or converges or static_high"
```

## Dependencies

- Related to F-ENG-01 (norm convention): the exact loss values will differ after #12 is fixed. The threshold-based assertions should use ratios (`loss[-1] > loss[0] * 10`, `loss[-1] < loss[0]`) rather than absolute values to remain robust.

## Risk

Low. These are pure additions; they do not change existing behavior.

**Comment by sachncs** [comment]
Resolved in PR #40.
